csv/src/fs2/data/csv/internals/CsvRowParser.scala

```diff
@@ -19,23 +19,22 @@ package csv
 package internals
 
 import cats._
+import cats.implicits._
 import cats.data._
 
 private[csv] object CsvRowParser {
 
-  def pipe[F[_], Header](withHeaders: Boolean)(
-      implicit F: ApplicativeError[F, Throwable],
-      Header: ParseableHeader[Header]): Pipe[F, NonEmptyList[String], CsvRow[Header]] =
-    _.mapAccumulate(Headers.Uninitialized[Header]: Headers[Header]) {
-      case (Headers.Uninitialized(), fields) if withHeaders =>
-        // headers have not been seen yet (first row)
-        (Headers.Initialized(Some(fields.map(Header.parse(_)))), None)
-      case (Headers.Uninitialized(), fields) =>
-        // no header are to be parsed
-        (Headers.Initialized(None), Some(new CsvRow[Header](fields, None)))
-      case (initialized @ Headers.Initialized(headers), fields) =>
-        // otherwise, headers are already initialized properly, just pass them to the row
-        (initialized, Some(new CsvRow[Header](fields, headers)))
-    }.map(_._2).unNone
+  def pipe[F[_], Header](implicit F: ApplicativeError[F, Throwable],
+                         Header: ParseableHeader[Header]): Pipe[F, NonEmptyList[String], CsvRow[Header]] =
+    _.pull.uncons1.flatMap {
+      case Some((firstRow, tail)) =>
+        Header(firstRow) match {
+          case Left(error) => Pull.raiseError[F](error)
+          case Right(headers) if headers.length =!= firstRow.length =>
+            Pull.raiseError[F](new HeaderError("The count of headers must match the count of columns"))
```

Comment:
Potentially add the sizes in the message, I think it can be useful when this happens, to debug